src/EditorFeatures/CSharp/EventHookup/EventHookupCommandHandler_TabKeyCommand.cs

```diff
@@ -85,116 +79,181 @@ private void HandleTabWorker(ITextView textView, ITextBuffer subjectBuffer, Acti
             }
         }
 
-        // Blocking wait (if necessary) to determine whether to consume the tab and
-        // generate the event handler.
-        EventHookupSessionManager.CurrentSession.GetEventNameTask.Wait(cancellationToken);
+        var subjectBuffer = args.SubjectBuffer;
+        var caretPoint = args.TextView.GetCaretPoint(subjectBuffer);
+        if (caretPoint is null)
+            return false;
+
+        var currentSnapshot = subjectBuffer.CurrentSnapshot;
+        var document = currentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+        if (document is null)
+            return false;
+
+        // Now emit the event asynchronously.
+        var token = _asyncListener.BeginAsyncOperation(nameof(ExecuteCommand));
```

Comment:
from this point on, we're doing all our work asynchronously, with a BWI in effect.